DM-35947: Implement live obscore table updates in daf_butler

andy-slac opened the pull request:
Adds new registry manager class responsible for obscore table updates. 

## Checklist

- [x] ran Jenkins
- [ ] added a release note for user-visible changes to `doc/changes`

andy-slac committed:
Add obscore manager.

The new manager adds one more table to registry for storing obscore records.
Registry methods which insert new datasets call this manager to update
the table with new records.

andy-slac committed:
Avoid leaking temp folder in test_cliCmdQueryDatasets

andy-slac committed:
Add support for associate and certify methods

andy-slac committed:
Add support for exposure regions.

Exposure records do not include region, to get region we match exposure
with a visit (any visit) and use region from that visit. It only works
if a visit is already defined when per-exposure datasets are inserted.
This does not work for raw datasets, which are inserted before the visits
are defined based on the same raws.

andy-slac committed:
Re-factor record factory to make it more reusable.

TallJimbo commented on python/lsst/daf/butler/registry/obscore/_records.py:
```diff
@@ -0,0 +1,247 @@
+        Returns
+        -------
+        record : `dict` [ `str`, `Any` ]
+            ObsCore record represented a dictionary.
```
Type annotation and code say this could also return `None`.

TallJimbo commented on python/lsst/daf/butler/registry/obscore/_records.py:
```diff
@@ -0,0 +1,247 @@
+            """Return records of all chained collections."""
+            for collection_record in self.collections.getParentChains(key):
+                yield collection_record
+                yield from _all_collections(collection_record.key)
```
I'd certainly prefer not to have logic like this run per-row, as it seems to be doing.  I think whenever we insert new datasets they're grouped by RUN in advance, and of course we also get that kind of grouping when associating into TAGGED collections.  It'd be nice to restructure to take advantage of that.

andy-slac replied:
I have updated that logic (it has actually moved to _manager.py in later commit) to reduce the number of checks.

andy-slac committed:
Improve collection check logic to reduce number of calls

andy-slac committed:
Re-mplement collection handling in obscore manager.

Now obscore manager only supports either run collections specified via
regex (multiple), or a single tagged collection. Configuration class has
an additional attribute for the type of collections.

andy-slac committed:
Improve configuration for extra columns.

The extra_columns is now configured via a special pydantic class, which
simplifies validation.

Also add support for configuration of indices in obscore table.

TallJimbo commented on python/lsst/daf/butler/registry/interfaces/_obscore.py:
```diff
@@ -0,0 +1,163 @@
+
+    def add_datasets(self, refs: Iterable[DatasetRef]) -> None:
+        """Possibly add datasets to the obscore table, this method should be
+        called when new datasets are added to a RUN collection.
```
Probably best to make the second clause here a new sentence on a separate line.  Same for similar docstrings of other methods in this class.

TallJimbo commented on python/lsst/daf/butler/registry/interfaces/_obscore.py:
```diff
@@ -0,0 +1,163 @@
+
+    def disassociate(self, refs: Iterable[DatasetRef], collection: Optional[str] = None) -> None:
+        """Possibly remove datasets from the obscore table, this method should
+        be called when datasets are disassociated from a TAGGED collection.
```
(Putting a big-picture comment here just because it's a semi-related place to start a thread.)

It occurs to me that we might be able to handle the TAGGED-collection removal problem more robustly with `ON DELETE CASCADE` _if_ we were to create a separate ObsCore table for each `dataset_tags_*` table, and then defined the full ObsCore "table" as an actual SQL `UNION` `VIEW`.  We could also have separate ObsCore tables for RUN-watching vs. TAGGED-watching, and the `UNION` would take care of the potential for duplicates there and let them coexist in the same repo.  I think that might be a simple enough view in practice (there aren't _that_ many distinct sets of dimensions in the dataset types of interest to ObsCore, I think) to avoid confusing the query optimizer too much, in contrast to the rejected "make it all a view" approach.

But it'd of course be a big change from what you have here, and not something I'd advise doing on this ticket, even if you like the idea enough to try it before turning this on in a major repo.  I'm pretty sure it'd also require some new `DatasetRecordStorageManager` methods to provide access to those foreign keys.

andy-slac replied:
I thought about it but decided to keep things simple. Having multiple table also means that obscore manager would need an additional hook for when a new dataset type is created to also create those dynamic tables for it, and also re-define a view to include this new table. Optimizer is also a concern, I do not have much experience with views and it is hard to guess for me whether a simple view with multiple UNIONs will work OK for all queries. I suspect there maybe cases when optimization breaks for some non-trivial queries. I can try to experiment, but I'd need a working spatial indexing first.

TallJimbo replied:
I'm happy with deferring this indefinitely if you don't think it's an obvious win, and what we have here meets our needs for now.

TallJimbo commented on python/lsst/daf/butler/registry/interfaces/_obscore.py:
```diff
@@ -0,0 +1,163 @@
+        """
+        raise NotImplementedError()
+
+    def associate(self, refs: Iterable[DatasetRef], collection: Optional[str] = None) -> None:
```
I think it's a bit more common for `Manager` interfaces to take `CollectionRecord` objects directly, and leave it to the calling `Registry` code to lookup records from names; that tends to avoid duplicate lookups better.  And it can sometimes save one manager from having to hold an instance of another, which I think is good when it works.

andy-slac replied:
For these methods I only need the name of a collection to match it against the configured tagged collection name. But I can make them to accept a CollectionRecord, just in case it will need to do something more complicated.

TallJimbo replied:
Ah, if you don't actually need it for anything else then I don't have a strong opinion on this.

TallJimbo commented on python/lsst/daf/butler/registry/managers.py:
```diff
@@ -128,7 +140,37 @@ def fromConfig(cls, config: RegistryConfig) -> RegistryManagerTypes:
+                try:
+                    class_name = manager_config_dict.pop("cls")
+                except KeyError:
+                    raise KeyError(f"'cls' key is not defined in {manager!r} manager configuration")
```
```suggestion
                    raise KeyError(f"'cls' key is not defined in {manager!r} manager configuration") from None
```
This will avoid those noisy "while one exception was being handled..." extra tracebacks.

andy-slac replied:
👍 Old habits, I still keep forgetting that I need `from` for exceptions.

TallJimbo commented on python/lsst/daf/butler/registry/managers.py:
```diff
@@ -204,6 +249,16 @@ def loadRepo(self, database: Database) -> RegistryManagerInstances:
+            # Get ObsCore configuration from attributes table, this silently
+            # overrides whatever may come from config file. Idea is that we do
+            # not want to carry around the whole thing, and butler config will
+            # have empty obscore configuration after initialization.
```
If someone creates a repo with non-empty obscore configuration, do we actually have logic to make that result in empty configuration in the repo's `butler.yaml` file?

andy-slac replied:
We don't have that logic yet, I'll try to find a place where to add it.

andy-slac replied:
I implemented stripping of obscore config before writing `butler.yaml` on [8305317](https://github.com/lsst/daf_butler/pull/732/commits/8305317fbc57c5fb4d15c975f087fcd92fa7ebc1). I'm not super happy about it because it adds some knowledge of obscore to Butler class, but I could not figure out a better place for that logic.

TallJimbo commented on python/lsst/daf/butler/registry/obscore/_manager.py:
```diff
@@ -0,0 +1,358 @@
+                },
+                universe=self.universe,
+            )
+            records = visit_detector_region_storage.fetch(DataCoordinateIterable.fromScalar(visitDataId))
```
It's really not ideal to be doing these fetches one data ID at a time.  Any chance you could at least rework the internal interfaces to support vectorized lookups, even if the implementation remains per-data ID fetches?  I think some changes coming on DM-31725 might make it easier to actually provide a vectorized implementation, so I'm fine with deferring that.

TallJimbo replied:
Actually, it looks like there's now a totally different approach to this in `dax_obscore`, and that's already able to handle lots of datasets at once.  Is this just dead code here?

andy-slac replied:
Would be nice indeed to have a vectorized `fetch`, I'll look at it when DM-31725 is done.

`dax_obscore` fetches too many records as an optimization in the assumption that it is doing bulk conversion and will need a large fraction of those records (which may not be true). That comes with a significant memory overhead which is probably fine for a standalone tool, but I would not want to do the same thing in a generic case. In the case of a regular ingest of raws (this is probably the only dataset that is going to use this logic) there won't be visit defined yet, so this should normally result in 0 records. In this case it will return early after trying to find a visit definition few lines above.

TallJimbo commented on python/lsst/daf/butler/registry/obscore/_manager.py:
```diff
@@ -0,0 +1,358 @@
+            config=config,
+            dimensions=dimensions,
+        )
+        self.tagged_collection = tagged_collection
```
I wonder if it might be better to have one concrete `ObsCoreLiveTableManager` implementation with a `collection_type` parameter.  The penalty for splitting it into three is just boilerplate, which isn't so bad, but it is a lot of boilerplate compared to what seems like it'd just be a few `if` statements.  And if we did someday work out a way to make RUN-watching and TAGGED-watching coexist in a single repo, a single manager class would be more future-proof.

andy-slac replied:
I started with that idea, but then decided to split it into two classes to avoid that mixed/exclusive state and bunch of `if`s and annoying `mypy` asserts. I can put it back together, should not be too hard.

TallJimbo replied:
If you've already tried it, please stick with this, because you already know better than I do how the alternative would look.

TallJimbo commented on tests/test_obscore.py:
```diff
@@ -0,0 +1,468 @@
+        self.assertEqual(len(rows), 0)
+
+
+class SQliteObsCoreTest(ObsCoreTests, unittest.TestCase):
```
```suggestion
class SqliteObsCoreTest(ObsCoreTests, unittest.TestCase):
```
or `SQLite`, if you prefer.  Just not `SQlite` :slightly_smiling_face: .

timj committed:
Remove mypy type ignores that v0.981 no longer needs

Mostly these are associated with cached getter properties.

andy-slac committed:
Fix for DimensionConfig construction in SqlRegistry

andy-slac committed:
Strip obscore configuration before writing butler.yaml
